File: examples/python/transactions.py
"""
OverDrive-DB — Transactions Example (Python)

Demonstrates: callback pattern, manual pattern, retry helper, isolation levels.

Install:
    pip install overdrive-db

Run:
    python transactions.py
"""
from overdrive import OverDrive, TransactionError

def main():
    db = OverDrive.open("transactions.odb")

    # Seed some data
    db.insert("accounts", {"id": "acc_alice", "owner": "Alice", "balance": 1000})
    db.insert("accounts", {"id": "acc_bob",   "owner": "Bob",   "balance": 500})
    print("Initial balances:")
    for acc in db.findAll("accounts"):
        print(f"  {acc['owner']}: ${acc['balance']}")

    # ── Callback Pattern (v1.4 — recommended) ──────────────────────────────
    # The callback receives the transaction ID.
    # On success → auto-commit. On exception → auto-rollback.
    print("\n=== Callback Transaction ===")

    def transfer_funds(txn):
        db.updateMany("accounts", "id = 'acc_alice'", {"balance": 900})
        db.updateMany("accounts", "id = 'acc_bob'",   {"balance": 600})
        return "transferred $100 from Alice to Bob"

    result = db.transaction(transfer_funds)
    print(f"Result: {result}")

    # ── Callback with return value ──────────────────────────────────────────
    print("\n=== Transaction with Return Value ===")

    def create_order(txn):
        order_id = db.insert("orders", {"item": "widget", "qty": 3, "status": "pending"})
        db.insert("audit_log", {"event": "order_created", "order_id": order_id})
        return order_id

    order_id = db.transaction(create_order)
    print(f"Created order: {order_id}")

    # ── Rollback on Exception ───────────────────────────────────────────────
    print("\n=== Automatic Rollback on Error ===")
    count_before = db.count("orders")

    try:
        def failing_operation(txn):
            db.insert("orders", {"item": "gadget", "qty": 1})
            raise ValueError("Something went wrong — transaction will be rolled back")

        db.transaction(failing_operation)
    except ValueError as e:
        print(f"Caught expected error: {e}")

    count_after = db.count("orders")
    print(f"Orders before: {count_before}, after: {count_after} (unchanged — rolled back)")

    # ── Retry Helper ────────────────────────────────────────────────────────
    # transaction_with_retry() retries on TransactionError (e.g. deadlock)
    # with exponential backoff. Other exceptions propagate immediately.
    print("\n=== Transaction with Retry ===")

    def safe_insert(txn):
        db.insert("events", {"type": "page_view", "url": "/home"})
        return "event logged"

    result = db.transaction_with_retry(safe_insert, max_retries=3)
    print(f"Retry result: {result}")

    # ── Isolation Levels ────────────────────────────────────────────────────
    print("\n=== Isolation Levels ===")

    # READ_COMMITTED (default) — no dirty reads
    def read_committed_txn(txn):
        return db.findAll("accounts")

    accounts = db.transaction(read_committed_txn, isolation=OverDrive.READ_COMMITTED)
    print(f"READ_COMMITTED: {len(accounts)} accounts")

    # SERIALIZABLE — full serial isolation
    def serializable_txn(txn):
        count = db.countWhere("accounts", "balance > 0")
        db.insert("snapshots", {"account_count": count})
        return count

    count = db.transaction(serializable_txn, isolation=OverDrive.SERIALIZABLE)
    print(f"SERIALIZABLE: {count} accounts with positive balance")

    # ── Manual Pattern (v1.3 — still fully supported) ───────────────────────
    print("\n=== Manual Transaction (v1.3 style) ===")
    txn_id = db.begin_transaction()
    try:
        db.insert("logs", {"event": "manual_txn_test", "status": "started"})
        db.commit_transaction(txn_id)
        print("Manual transaction committed")
    except Exception as e:
        db.abort_transaction(txn_id)
        print(f"Manual transaction rolled back: {e}")

    # ── Context Manager Pattern ─────────────────────────────────────────────
    print("\n=== Context Manager Pattern ===")
    with db.transaction() as txn_id:
        db.insert("logs", {"event": "context_manager_txn", "txn_id": txn_id})
    print("Context manager transaction committed")

    # Final state
    print("\n=== Final Balances ===")
    for acc in db.findAll("accounts"):
        print(f"  {acc['owner']}: ${acc['balance']}")

    db.close()
    print("\nDone!")

if __name__ == "__main__":
    main()

File: examples/nodejs/transactions.js
/**
 * OverDrive-DB — Transactions Example (Node.js)
 *
 * Demonstrates: callback pattern (sync and async), manual pattern,
 * retry helper, and isolation levels.
 *
 * Install:
 *   npm install overdrive-db
 *
 * Run:
 *   node transactions.js
 */
const { OverDrive, TransactionError } = require('overdrive-db');

async function main() {
    const db = OverDrive.open('transactions.odb');

    // Seed some data
    db.insert('accounts', { id: 'acc_alice', owner: 'Alice', balance: 1000 });
    db.insert('accounts', { id: 'acc_bob',   owner: 'Bob',   balance: 500 });
    console.log('Initial balances:');
    db.findAll('accounts').forEach(acc => console.log(`  ${acc.owner}: $${acc.balance}`));

    // ── Callback Pattern (v1.4 — recommended) ──────────────────────────────
    // The callback receives the transaction ID.
    // On success → auto-commit. On exception → auto-rollback.
    console.log('\n=== Callback Transaction (sync) ===');

    const result = db.transaction((txn) => {
        db.updateMany('accounts', "id = 'acc_alice'", { balance: 900 });
        db.updateMany('accounts', "id = 'acc_bob'",   { balance: 600 });
        return 'transferred $100 from Alice to Bob';
    });
    console.log('Result:', result);

    // ── Async Callback ──────────────────────────────────────────────────────
    // transaction() supports async callbacks — returns a Promise.
    console.log('\n=== Async Callback Transaction ===');

    const asyncResult = await db.transaction(async (txn) => {
        const orderId = db.insert('orders', { item: 'widget', qty: 3, status: 'pending' });
        db.insert('audit_log', { event: 'order_created', orderId });
        // Simulate async work (e.g. calling an external service)
        await new Promise(resolve => setTimeout(resolve, 10));
        return orderId;
    });
    console.log('Created order:', asyncResult);

    // ── Rollback on Exception ───────────────────────────────────────────────
    console.log('\n=== Automatic Rollback on Error ===');
    const countBefore = db.count('orders');

    try {
        await db.transaction(async (txn) => {
            db.insert('orders', { item: 'gadget', qty: 1 });
            throw new Error('Something went wrong — transaction will be rolled back');
        });
    } catch (e) {
        console.log('Caught expected error:', e.message);
    }

    const countAfter = db.count('orders');
    console.log(`Orders before: ${countBefore}, after: ${countAfter} (unchanged — rolled back)`);

    // ── Retry Helper ────────────────────────────────────────────────────────
    // transactionWithRetry() retries on TransactionError (e.g. deadlock)
    // with exponential backoff. Other exceptions propagate immediately.
    console.log('\n=== Transaction with Retry ===');

    const retryResult = await db.transactionWithRetry(async (txn) => {
        db.insert('events', { type: 'page_view', url: '/home' });
        return 'event logged';
    }, OverDrive.READ_COMMITTED, 3);
    console.log('Retry result:', retryResult);

    // ── Isolation Levels ────────────────────────────────────────────────────
    console.log('\n=== Isolation Levels ===');

    // READ_COMMITTED (default) — no dirty reads
    const accounts = db.transaction(
        (txn) => db.findAll('accounts'),
        OverDrive.READ_COMMITTED
    );
    console.log('READ_COMMITTED:', accounts.length, 'accounts');

    // SERIALIZABLE — full serial isolation
    const snapshotCount = db.transaction((txn) => {
        const count = db.countWhere('accounts', 'balance > 0');
        db.insert('snapshots', { accountCount: count });
        return count;
    }, OverDrive.SERIALIZABLE);
    console.log('SERIALIZABLE:', snapshotCount, 'accounts with positive balance');

    // ── Manual Pattern (v1.3 — still fully supported) ───────────────────────
    console.log('\n=== Manual Transaction (v1.3 style) ===');
    const txnId = db.beginTransaction();
    try {
        db.insert('logs', { event: 'manual_txn_test', status: 'started' });
        db.commitTransaction(txnId);
        console.log('Manual transaction committed');
    } catch (e) {
        db.abortTransaction(txnId);
        console.log('Manual transaction rolled back:', e.message);
    }

    // Final state
    console.log('\n=== Final Balances ===');
    db.findAll('accounts').forEach(acc => console.log(`  ${acc.owner}: $${acc.balance}`));

    db.close();
    console.log('\nDone!');
}

main().catch(err => {
    console.error('Error:', err.message);
    process.exit(1);
});
